### PR TITLE
Add VictorOps Receiver type to AlertmanagerConfig CRD

### LINKED ISSUE #3539: Add VictorOps Receiver type to AlertmanagerConfig CRD
**What is missing?**

In #3451 we added the AlertmanagerConfig CRD, with just the PagerDuty receiver type, meaning that alert notifications can only be routed to PagerDuty. Alertmanager supports several other Receiver types, with this being one (other issues will be created for the others).

**Why do we need it?**

We should be able to define all of the Alertmanager receivers in AlertmanagerConfig CRs.

**Anything else we need to know?**:

The proposal for AlertmanagerConfig CRD is [here](https://docs.google.com/document/d/1aVVttvocop8zNezwrNFbHh_HKRW_fhJUZqhmrr8fGK0/edit#heading=h.5ua2yyj2xm4h), and it includes some information on this receiver type.

The Alertmanager config docs [here](https://prometheus.io/docs/alerting/latest/configuration/) show how this receiver is expected to be in the resulting Alertmanager config.

### PR DESCRIPTION
**Pull request by sev3ryn**
fixes #3539

### CODE REVIEW

**Comment by simonpasquier** [comment]
Thanks! cc @grdryn

File: pkg/alertmanager/types.go [hunk 1]
```diff
@@ -279,14 +279,14 @@ func (d duration) MarshalText() ([]byte, error) {
...
-	MessageType       string                      `yaml:"message_type" json:"message_type"`
-	StateMessage      string                      `yaml:"state_message" json:"state_message"`
-	EntityDisplayName string                      `yaml:"entity_display_name" json:"entity_display_name"`
-	MonitoringTool    string                      `yaml:"monitoring_tool" json:"monitoring_tool"`
-	CustomFields      map[string]string           `yaml:"custom_fields,omitempty" json:"custom_fields,omitempty"`
+	VSendResolved     bool              `yaml:"send_resolved" json:"send_resolved"`
+	HTTPConfig        *httpClientConfig `yaml:"http_config,omitempty" json:"http_config,omitempty"`
+	APIKey            string            `yaml:"api_key,omitempty" json:"api_key,omitempty"`
+	APIURL            string            `yaml:"api_url,omitempty" json:"api_url,omitempty"`
+	RoutingKey        string            `yaml:"routing_key,omitempty" json:"routing_key,omitempty"`
```
Comment (grdryn) [inline pkg/alertmanager/types.go:286]: It seems like at least the `routingKey` field here shouldn't be allowed to be optional, I'm not sure about the others. Thoughts?

See here where Alertmanager considers an empty/missing `routingKey` an error: https://github.com/prometheus/alertmanager/blob/master/config/notifiers.go#L510..L512

Comment (sev3ryn) [reply pkg/alertmanager/types.go:286]: nice catch - fixed in 2b59a11

**Comment by lilic** [comment]
@sev3ryn seems like this needs another rebase. :) Thanks for the hard work and patience!

**Comment by sev3ryn** [comment]
@lilic  I'll rebase this first to be merged and #3697 after this will be merged to master as it will also have conflicts

**Comment by lilic** [comment]
Sounds like a good plan!

**Comment by sev3ryn** [comment]
@lilic rebased and also fixed example in documentation - see 6f760d3c

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 2]
```diff
@@ -122,6 +122,8 @@ type Receiver struct {
 	WeChatConfigs []WeChatConfig `json:"wechatConfigs,omitempty"`
 	// List of Email configurations.
 	EmailConfigs []EmailConfig `json:"emailConfigs,omitempty"`
+	// List of VictorOps configurations.
+	VictorOpsConfigs []VictorOpsConfig `json:"victorOpsConfigs,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:126]: To be compliant with #3690 

```suggestion
	VictorOpsConfigs []VictorOpsConfig `json:"victoropsConfigs,omitempty"`
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 3]
```diff
@@ -444,6 +446,29 @@ type EmailConfig struct {
...
+// See https://prometheus.io/docs/alerting/latest/configuration/#victorops_config
+type VictorOpsConfig struct {
+	// Whether or not to notify about resolved alerts.
+	SendResolved *bool `json:"sendResolved,omitempty"`
+	// The API key to use when talking to the VictorOps API.
+	APIKey *v1.SecretKeySelector `json:"apiKey,omitempty"`
+	// The VictorOps API URL.
+	APIURL *string `json:"apiUrl,omitempty"`
+	// A key used to map the alert to a team.
+	RoutingKey *string `json:"routingKey,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:459]:
```suggestion
	RoutingKey *string `json:"routingKey"`
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 4]
```diff
@@ -444,6 +446,29 @@ type EmailConfig struct {
 	TLSConfig *monitoringv1.SafeTLSConfig `json:"tlsConfig,omitempty"`
 }
 
+// VictorOpsConfig configures notifications via VictorOps.
+// See https://prometheus.io/docs/alerting/latest/configuration/#victorops_config
+type VictorOpsConfig struct {
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:451]: Don't we miss `CustomFields` here? This should be `[]KeyValue`.

**Review by simonpasquier — COMMENTED** [review]
Nice :) A few remarks only.

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go (hunk 4 above)
Comment (sev3ryn) [reply pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:451]: I don't see CustomFields is alertmanager configuration
https://prometheus.io/docs/alerting/latest/configuration/#victorops_config

if it is planned in future alertmanager release I believe it should be added later. But anyways if you want it to be added I'll add

Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:451]: I should have added the link to the code in the first place :)
https://github.com/prometheus/alertmanager/blob/a7f9fdadbecbb7e692d2cd8d3334e3d6de1602e1/config/notifiers.go#L483

The documentation is out-of-date unfortunately.

Comment (sev3ryn) [reply pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:451]: fixed in 348657e6bdd3d62f75a53bccd26686576ba5155d.

**Comment by sev3ryn** [comment]
before merging this PR - please let me know - so I'll rebase all the fixup commits - to make git history nicer :)

**Comment by simonpasquier** [comment]
thanks for the quick update! We can squash all the commits before merging so no real need for you to autosquash things.

**Comment by simonpasquier** [comment]
lgtm

**Pull request merged**